### PR TITLE
perf: add React.lazy to the Routes

### PR DESCRIPTION
**Pull request by HowardBraham**
## **Description**

> **Note:** I would really appreciate quick reviews here mostly based on "did I break anything."  It's not really worth our time to nit-pick about the style or the details because every line of this is going to be changed by me again in the near future.  Still to do in later PRs that are going to completely re-write what I did here:
> - Convert routes.component.js to TypeScript
> - Convert routes.component.js to Hooks
> - Convert routes.component.js to react-router v6
> - Eliminate routes.container.js
>
> All this has to be done in sequential PRs based on top of this one, so while I wait for reviews I can only work on a non-Routes topic.

This is the first PR to add `React.lazy` to the codebase.  The meat of the changes are in `ui/pages/routes/routes.component.js`.

Every other file changed is just a reaction to those changes.  Much of this is caused by the fact that `React.lazy` only works on default exports ([see the documentation here](https://react.dev/reference/react/lazy)), so I had to change a few named exports to default exports.

I don't think this PR has much of an impact on loading times, but it sets up further work that should have an impact.

There was one component (`ConfirmTransaction`) that needed some extra attention to work with `React.lazy`.  I had to remove the `history.replace(mostRecentOverviewPage);`, which I'm pretty sure was an unnecessary statement, and remove the Unit Test related to it. 

**UPDATE:** I changed all of the Integration Tests to convert `getBy...` to `await findBy...` to wait for the `React.lazy` component to load.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/28172?quickstart=1)

## **Related issues**

Progresses: MetaMask/MetaMask-planning#3302

### CODE REVIEW

File: ui/components/multichain/pages/connections/index.js [hunk 1]
```diff
@@ -1 +1 @@
-export { Connections } from './connections';
```
Comment (matthewwalsh0) [inline ui/components/multichain/pages/connections/index.js:1]: I hit this in my POC also, is there definitely no workaround for this as we tend to prefer non-default exports for auto-complete and extensibility?

File: ui/pages/routes/routes.component.js [hunk 2]
```diff
@@ -122,6 +96,66 @@ import {
   showOnboardingHeader,
 } from './utils';
 
+// Begin Lazy Routes
+const OnboardingFlow = React.lazy(() =>
```
Comment (matthewwalsh0) [inline ui/pages/routes/routes.component.js:100]: Given the volume of changes here and limited readability, is it worth listing these in alphabetical order?

File: ui/pages/routes/routes.component.js [hunk 3]
```diff
@@ -122,6 +96,66 @@ import {
...
+  import('../onboarding-flow/onboarding-flow'),
+);
+const Lock = React.lazy(() => import('../lock'));
```
Comment (matthewwalsh0) [inline ui/pages/routes/routes.component.js:103]: Minor, but something I did in my POC was create a helper function for this so we had an easy injection point for logging and timing.

https://github.com/MetaMask/metamask-extension/compare/develop...poc/react-lazy-load#diff-2466ed85b744b657eb0b324376eef5486080bf2b897eaa3af9b8c3ed0184ebeaR117

File: ui/pages/routes/routes.component.js [hunk 4]
```diff
@@ -122,6 +96,66 @@ import {
...
+const Settings = React.lazy(() => import('../settings'));
+const NotificationsSettings = React.lazy(() =>
+  import('../notifications-settings'),
+);
+const NotificationDetails = React.lazy(() => import('../notification-details'));
+const Notifications = React.lazy(() => import('../notifications'));
+const SnapList = React.lazy(() => import('../snaps/snaps-list'));
+const SnapView = React.lazy(() => import('../snaps/snap-view'));
+
+// eslint-disable-next-line -- Un-lazy this one for now
```
Comment (matthewwalsh0) [inline ui/pages/routes/routes.component.js:118]: Perhaps more readable to have an alphabetical list of lazy routes, and then an alphabetical list of standard routes defined with comments?

File: ui/pages/routes/routes.component.js [hunk 5]
```diff
@@ -260,115 +294,120 @@ export default class Routes extends Component {
...
-        />
-        <Authenticated path={PERMISSIONS} component={PermissionsPage} exact />
-        <Authenticated
-          path={`${REVIEW_PERMISSIONS}/:origin`}
-          component={ReviewPermissions}
-          exact
-        />
-        <Authenticated path={DEFAULT_ROUTE} component={Home} />
-      </Switch>
+      <Suspense fallback={<AppLoadingSpinner />}>
```
Comment (matthewwalsh0) [inline ui/pages/routes/routes.component.js:297]: Has this been approved by the design team, does this match the initial loader on first load?

**Review by matthewwalsh0 — DISMISSED** [review]

File: ui/components/multichain/pages/connections/index.js (hunk 1 above)
Comment (HowardBraham) [reply ui/components/multichain/pages/connections/index.js:1]: Okay I found this method and it seems to work: https://dev.to/iamandrewluca/react-lazy-without-default-export-4b65

So we would write it as
```
const Connections = React.lazy(() =>
  import('../../components/multichain/pages/connections').then((module) => ({
    default: module.Connections,
  })),
);
```

But I'm not totally convinced we should do that.  For one thing, it's kind of hard to follow.  And second, people say it may limit tree-shaking.

File: ui/pages/routes/routes.component.js (hunk 2 above)
Comment (HowardBraham) [reply ui/pages/routes/routes.component.js:100]: Hah, I guess it looked a little chaotic to you, but actually I was following an order.  The lazy route imports are in the same order as the `<Switch>` component.

File: ui/pages/routes/routes.component.js (hunk 4 above)
Comment (HowardBraham) [reply ui/pages/routes/routes.component.js:118]: Yeah this is in the middle of the list to keep the same order as the `<Switch>` component.  Eventually I hope to figure it out and make it lazy again.

File: ui/pages/routes/routes.component.js (hunk 5 above)
Comment (HowardBraham) [reply ui/pages/routes/routes.component.js:297]: Oh great point, the load is so fast that I didn't even look at it.  But once I slowed it down on purpose, actually you can't even see the spinner.  The extension becomes Header only, and the rest of the page just isn't loaded yet.

Yeah I wonder if it should actually match the orange spinner with the fox, or the gray spinner.  Lemme ask the design team.

File: ui/pages/routes/routes.component.js (hunk 3 above)
Comment (HowardBraham) [reply ui/pages/routes/routes.component.js:103]: This is a fascinating thought, and maybe the helper function could even handle the non-default-export thing.

**Review by Joj501 — APPROVED** [review]

**Review by Joj501 — APPROVED** [review]
Approved change

**Review by pedronfigueiredo — APPROVED** [review]

**Review by jonybur — CHANGES_REQUESTED** [review]

**Review by jonybur — APPROVED** [review]
